### DISCUSSION TITLE
streamkmeans visualization

### DISCUSSION THREAD

**Discussion by drakeleeb** [online-ml/river#899] [Q&A]
How do I pull information to plot as streamkeans is running? Not sure where is the best place to grab the evolving cluster centers and clustered data. Are there any examples for this use-case? Thanks.

**Comment by MaxHalford** [comment]
@smastelini @hoanganhngo610 have fun with this one :) (seems there's a lot of interest in clustering!). Also @vsoch might want to chime in. I'm sure it's possible to do some matplotlib judo to get this to work. We just don't have any good examples right now in the documentation, as far as I'm aware of.

In general, I don't think it should be River's goal to provide visualization utilities. There are just too many use cases to cater to. We could however imagine providing another library for online viz purposes. It's just a lot of work right now.

**Reply by vsoch** [reply]
@drakeleeb I've done something similar! It could be an "on demand" view (given you have a web interface that someone will load and then run it on demand) or you could have it run at some interval / by some other trigger. For:

1. Determining if a model has centers and finding them
2. Sending the data to a web interface to render
3. (optionally) finding the currently assigned data

The general steps I used are the following:

Get the model and make sure it has centroids (not all do, and maybe you can skip this step if you are only using clustering models: https://github.com/spack/spack-monitor/blob/ad70f9fce241cf866ee10ebddd32564dc18b5c00/spackmon/apps/main/views/analysis.py#L26-L43

```python
@ratelimit(key="ip", rate=rl_rate, block=rl_block)
@never_cache
def view_clusters(request):
    """
    View machine learning clusters for our model!
    """
    # Get a django client
    client = DjangoClient()

    model = client.get_model(cfg.MODEL_NAME)
    have_centroids = False
    if online_ml.get_centers(model) is not None:
        have_centroids = True
    return render(
        request,
        "online_ml/clusters.html",
        {"have_centroids": have_centroids, "model_name": cfg.MODEL_NAME},
    )
```
From a web UI, I want the centroids to be fetched by the view on demand (and not grabbed in advance and then passed to it) so I have a function that checks for centroids:

```python
if online_ml.get_centers(model) is not None:
    have_centroids = True
```
and that is basically looking at model.steps or directly at model.centers (you'll have model.steps for a pipeline, more common, and just the model with centers directly on it without any other steps): https://github.com/spack/spack-monitor/blob/ad70f9fce241cf866ee10ebddd32564dc18b5c00/spackmon/apps/main/online_ml.py#L22-L31

```python
def get_centers(model):
    """
    Helper function to derive centroids from a model
    """
    if hasattr(model, "steps"):
        for step_name, step in model.steps.items():
            if hasattr(step, "centers") and step.centers:
                return step.centers
    elif hasattr(model, "centers") and model.centers:
        return model.centers
```
Once you know that you have centroids, what actually I found easy to do is just pass the boolean to the interface, and have a request to a different endpoint explicitly to return the centroids (and then render into a visualization, for example): https://github.com/spack/spack-monitor/blob/ad70f9fce241cf866ee10ebddd32564dc18b5c00/spackmon/apps/main/templates/online_ml/clusters.html#L160-L177. That hits this endpoint function: https://github.com/spack/spack-monitor/blob/ad70f9fce241cf866ee10ebddd32564dc18b5c00/spackmon/apps/main/views/analysis.py#L46-L53

```python
def get_centroids(request, name):
    """
    JsonResponse to just retrun centroids
    """
    client = DjangoClient()
    model = client.get_model(name)
    centers = online_ml.get_centers(model)
    return JsonResponse({"centers": online_ml.generate_embeddings(centers)})
```
Which is going to use that same function to get the centers, but instead give them to another function to return a reduced version (e.g., high dimension reduced to 2 so we can plot x and y). I basically generate a distance matrix then run TSNE and return that. That happens here https://github.com/spack/spack-monitor/blob/ad70f9fce241cf866ee10ebddd32564dc18b5c00/spackmon/apps/main/online_ml.py#L43-L63

```python

def generate_embeddings(centers):
    df = pandas.DataFrame(centers)

    # 200 rows (centers) and N columns (words)
    df = df.transpose()
    df = df.fillna(0)

    # Create a distance matrix
    distance = pandas.DataFrame(
        squareform(pdist(df)), index=list(df.index), columns=list(df.index)
    )

    # Make the tsne (output embeddings go into docs for visual)
    fit = manifold.TSNE(n_components=2)
    embedding = fit.fit_transform(distance)
    emb = pandas.DataFrame(embedding, index=distance.index, columns=["x", "y"])
    emb.index.name = "name"

    # cluster ids
    emb["name"] = emb.index
    return emb.to_dict(orient="records")
```
But the above is a more complicated server setup - you can literally just grab the model centers, either at `model.steps["<model_name>"].centers` or just `model.centers` and do what you want with them! My particular approach generates this:

![image](https://user-images.githubusercontent.com/814322/160918508-0bb8a663-43b3-4768-9e5c-a5ac46c690dd.png)

And actually I just realized I added this example to Django River ML (A Django Plugin for river) in the sample all (how to run [here](https://vsoch.github.io/django-river-ml/getting_started/user-guide.html#sample-application)).

So that's a "do it yourself" means to get centroids and then view them, and if you wanted to actually assign your data to the current centroids you'd have to loop through some set of data and get a prediction (a number for a centroid) and then save that (and possibly allow a user to mouse over a centroid and then see assigned). A static version of that is here https://github.com/buildsi/spack-monitor-nlp-dev/blob/a99fa2fb780ad708ab1ee5b08cdeaa1222a2a83b/2.online-ml.py#L223-L247

But there are other cool ideas in your question:
- Can we have a library that provides visualization for river? I think so - the question would be what format do we want. For streaming what comes to mind (that makes sense) is some kind of simple flask app that uses sockets to update data at some frequency. Otherwise it would be "on demand" (e.g., refresh to see new updates). But choosing this approach would then force us to choose a particular server backend, and I don't think we'd want that - we'd want something that can be used in Flask or Django or whatever. Or we'd want to use some of the cool interactive libraries like bokeh / plotly / holoviz / pycharts. The choice really depends on your goal - a server approach is for server applications, and the others are more suited for data science. For a server that needs to load quickly, a main issue would be the size of the model / time to render a new update. And probably for the plotting /JS libraries we'd want to decide on that first, and make some simple examples.
- Can we keep track of data assignments (and update them) so we can (on demand) see some kind of visualization without needing to run the full set of predictions? This is actually the case that I ran into the other day https://github.com/online-ml/river/issues/891 and I concluded with changing clusters/microclusters (basically everything except for kmeans) it wouldn't be possible to store an assignment. But - if there is some way to represent state, e.g., after running a learn and having some change of cluster state:
  - "cluster 10 was split into 11 and 12 == reclassify just everything assigned with 10" 
  - "cluster 10 was deleted"  == reclassify everything that was in 10
  - "cluster 10 and 11 were merged in 12" == re-assign everything in 10 and 11 to 12.
Then you could always just use the current assignments of your datum (stored elsewhere, like your database) to plot quickly (with the centroids). The strength of river is that we don't save all the data points - but that's also (sort of) can be a weakness that the model loses memory over time. The assignments change as do the centroids. My [original efforts](https://buildsi.github.io/spack-monitor-nlp/) on my current dataset used Doc2Vec and embeddings, and it was great because I could plot *every single point* but then the data got larger and I could barely run anything. That's when I googled "streaming ML" or something similar and found river, haha.

> In general, I don't think it should be River's goal to provide visualization utilities. There are just too many use cases to cater to. We could however imagine providing another library for online viz purposes. It's just a lot of work right now.

Yep 100% agree - plotting support should be separate. I'd enjoy thinking / working on this if there is a larger interest and people want to discuss how to go about it! I suspect most folks at the moment are "rolling their own" and arguably it doesn't have to be that way.

**Reply by hoanganhngo610** [reply]
This, is just, basically, beyond any expectation that I can ever think of. To be honest, I myself cannot even think of anything more complete than this. Kudos to @vsoch, once again.

**Comment by hoanganhngo610** [comment]
@drakeleeb Thank you so much for raising such an interesting question. I think @vsoch have just covered most of what you need to know upon querying the cluster centers and the cluster results; however, I think that upon my way of doing clustering benchmarking, I found quite an interesting way to visualize the results live using `Holoviews`. In fact, this practice has been done by @jacobmontiel during his tutorial on River in IJCAI 2020. I am intending to write a short notebook for the clustering module itself, and it should be released within no time!

**Reply by vsoch** [reply]
@hoanganhngo610 what do you think of having some collection of river visualizations? E.g either a repository of notebooks for "on demand examples using whatever" or we could try to put our heads together and brainstorm a library framework - e.g., this last weekend I was using https://plotly.com/python/graph-objects/ and that will generate a more interactive webby UI, but then you can also easily save to file, e.g., https://raw.githubusercontent.com/vsoch/citelang/main/docs/assets/img/pypi-citelang.png and it was super easy! That would have taken me forever to figure out from scratch. So I wonder if we could have a visualization library that does basic functionality like grabbing centroids, doing dimensionality reduction? Here is my proposal I think:

1. Let's start with a repo of notebooks - and whatever examples people have, they can be added there. Easy "grab and go" kind of deal.
2. Then we can have a TODO list of the kinds of examples that we want
3. When we have enough examples / requests for examples we can get a sense of what 80% of people are looking for, and perhaps we will have tried a bunch of techniques for doing it. Then we can talk about designing a library that could basically wrap tools of interest. Something like:

```
# visualizations that require no additional deps - maybe just html/css sort of deal
pip install river-viz

# or other dependencies!
pip install river-viz[plotly]
pip install river-viz[bokeh]
```
the goal of the above being to not have dependency bloat or installing every viz library under the sun when you just want one.

**Reply by MaxHalford** [reply]
> Let's start with a repo of notebooks - and whatever examples people have, they can be added there. Easy "grab and go" kind of deal.

> When we have enough examples / requests for examples we can get a sense of what 80% of people are looking for, and perhaps we will have tried a bunch of techniques for doing it. Then we can talk about designing a library that could basically wrap tools of interest. Something like:

That's actually a good idea. Very pragmatic.

In general it sounds like a good idea to have this Python based. As you say, this is to satisfy 80% of needs, not be fully comprehensive. Something with plotly/streamlit or what not sounds good. But it's definitely worth exploring some options.

**Reply by smastelini** [reply]
I love that idea too! I'm happy to see a River ecosystem starting to grow!

**Reply by hoanganhngo610** [reply]
@vsoch I'm also rooting for this idea. Once again, my only concern is regarding the needs, but due to the emerging use of River within the recent time, I would not doubt that this would be of necessary need in the future. I would also agree that we can start with notebooks first to help people have some ideas on what are we trying to do/offer, then have something based on available libraries, for example plotly/streamlit.

**Reply by MaxHalford** [reply]
@vsoch I'm thinking it would be cool to have a gallery [like what Seaborn does](https://seaborn.pydata.org/examples/index.html). That way users can just copy/paste and adapt as they see fit. We can provide utilities to cover the repetitive stuff -- either a separate library, or a new `viz` module. 

Regarding libraries, I like what plotly does [here](https://github.com/plotly/Streaming-Demos).

**Reply by vsoch** [reply]
Sounds good! If you want to make a repo (or give me permission to) I can start a small UI. Do you want me to use the same template you have for the river docs, or something more akin to the Seaborn gallery (but branded for river?)

**Reply by vsoch** [reply]
there is also sphinx gallery, which is nice for Python and packaging examples with code in different formats (python file, notebooks, etc). https://sphinx-gallery.github.io/stable/index.html

**Reply by MaxHalford** [reply]
@vsoch I added you as a maintainer to [this](https://github.com/online-ml/viz) repo.

Personally I like the Seaborn gallery. I think the layout looks fantastic. A Sphinx gallery is also a good option, although I've had bad experiences with Sphinx. I really like [this](https://sphinx-gallery.github.io/stable/auto_examples/plot_8_animations.html#sphx-glr-auto-examples-plot-8-animations-py) example. The gallery could also live in the River documentation, which is based on MkDocs, so we have everything in a single place.

**Reply by drakeleeb** [reply]
I would like to track time series of the depth of ocean sills at several locations, cluster the time series, and update the clusters as they evolve over time. Would this be possible using River?

BTW, the idea to develop a repository of python notebooks for demonstration and  learning of River tools is an excellent idea!

**Reply by vsoch** [reply]
@drakeleeb that would be a great starting example for me - do you have a dummy dataset that I could use? Or a publicly available dataset already in river that could emulate what you have / are looking to do?

**Reply by vsoch** [reply]
okay got super distracted with other projects today but I started on something this evening - don't want to set an expectation but likely will get a base structure with a clear organization and "How to add a new example" online soon! I think I've found the right balance of how to do this:

- sphinx gallery with the right setup really gives us a lot of benefits - versioning, automated running of example code on a build, and a nice gallery functionality that is integrated with Python. As @MaxHalford showed above, this extends to animated stuffs too (which I think will be more important as we work on streaming sorts of examples).
- sphinx has a make the docs theme. We can pretty easily match the branding of river docs this way1 I should probably switch to using this theme for django-river-ml at some point.
- sphinx-gallery also is going to generate us all the kinds of examples we would want! As an early test, I set up a dummy gallery with a quick example to show finding centroids (which I showed above). Here is a quick and dirty "yay it shows up in the gallery!"

![image](https://user-images.githubusercontent.com/814322/161411261-08521609-9c96-47c9-ad39-01bbba89d61f.png)

I also tried with other visualization types to see the thumbnail  update - super cool!

![image](https://user-images.githubusercontent.com/814322/161410823-2e88818c-b4a6-4f21-87f6-f5e399206c8c.png)

And the full gallery example (I cropped out the sidebar / buttons on the bottom because they are super ugly right now, need to work on them a bit!)
![Screenshot 2022-04-02 at 22-17-20 Introductory example - Finding Model Data — River Visualizations](https://user-images.githubusercontent.com/814322/161411236-18c72805-1240-4f0a-90b1-a4cda8808f8b.png)


And here is me downloading an example and running it as a notebook:

![image](https://user-images.githubusercontent.com/814322/161410812-da6055ee-de09-4cad-9448-eb274c34e796.png)

So I think this will be an awesome start! I've used sphinx-gallery sporadically and I suspect I'll get a lot better at it using it here. I will ping y'all when I have something pushed / ready to contribute to / ready to make a viz requests list.

**Reply by vsoch** [reply]
Also if y'all have examples in notebooks / python anywhere that you would want to add, send them to me sooner than later - it's helpful to have a wide range while I'm thinking about the organization, etc! A lot of my time tonight was just writing out that example.

**Reply by MaxHalford** [reply]
Very cool! I think we have some decision tree plots that would nicely here. Also some progressive validation charts would be nice to nice. I can contribute them once this is set up. Let us know how we can help :)

**Reply by vsoch** [reply]
Will do! I should be able to work on it most of today (already started) and get something up as soon as I get the design / branding / layout / automation to my liking. Since it's sphinx we will want to decide to deploy to readthedocs (supports automated builds and easy versioning) versus manual to GitHub pages. I'm good with either - the first is just connecting the repo and setting up in readthedocs, and the second I can write a workflow that deploys to GitHub pages.

**Reply by MaxHalford** [reply]
Gut feeling is that GitHub Pages would be nice because the URL would be `riverml.xyz/viz`. That being said, if readthedocs allows using a custom domain, then it's all good

**Reply by vsoch** [reply]
@MaxHalford I am setting it up - I need more permissions than you've given me to enable head branch cleanup, add the CircleCI integration, and enable pages. I'm a pretty seasoned developer, you can trust me with these things.

**Reply by vsoch** [reply]
okay here is the start! https://online-ml.github.io/viz/ It looks like I do have permissions to enable pages (and @MaxHalford could you please enable branch cleanup?) When @MaxHalford approves I have a circle recipe ready to go to allow us to preview the site when someone contributes a PR. Most of my time this morning was tweaking the style/branding of the template, and then writing contributing docs. We're on our way! :tada:
